File: roborio/c++/navXMXP_CPlusPlus_RobotExample/src/Robot.cpp
#include "WPILib.h"
#include "IMU.h"
#include "IMUAdvanced.h"
#include "AHRS.h"

/**
 * This is a demo program showing the use of the RobotBase class.
 * The SimpleRobot class is the base of a robot application that will automatically call your
 * Autonomous and OperatorControl methods at the right time as controlled by the switches on
 * the driver station or the field controls.
 */
class Robot : public SampleRobot
{
	NetworkTable *table;
	AHRS *imu;
	SerialPort *serial_port;
	bool first_iteration;

public:
	Robot()
	{
	}

	void RobotInit() {

		table = NetworkTable::GetTable("datatable");
		serial_port = new SerialPort(57600,SerialPort::kMXP);
        uint8_t update_rate_hz = 50;
        imu = new AHRS(serial_port,update_rate_hz);
        if ( imu ) {
        	LiveWindow::GetInstance()->AddSensor("IMU", "Gyro", imu);
        }
        first_iteration = true;
	}

	/**
	 *
	 * Wait for 10 seconds
	 */
	void Autonomous(void)
	{
		Wait(2.0); 				//    for 10 seconds
	}

	/**
	 * Runs the motors with arcade steering.
	 */
	void OperatorControl(void)
	{
		int reset_yaw_count = 0;
		int reset_displacement_count = 0;
		while (IsOperatorControl())
		{
			if ( first_iteration ) {
	            bool is_calibrating = imu->IsCalibrating();
	            if ( !is_calibrating ) {
	                Wait( 0.3 );
	                imu->ZeroYaw();
	                first_iteration = false;
	            }
			}
            bool yaw_axis_up;
            uint8_t yaw_axis = imu->GetBoardYawAxis(yaw_axis_up);

            bool reset_yaw_button_pressed = DriverStation::GetInstance()->GetStickButton(0,1);
            if ( reset_yaw_button_pressed ) {
            	imu->ZeroYaw();
            	reset_yaw_count++;
            }
            bool reset_displacement_button_pressed = DriverStation::GetInstance()->GetStickButton(0,2);
            if ( reset_displacement_button_pressed ) {
            	imu->ResetDisplacement();
            	reset_displacement_count++;
            }

            SmartDashboard::PutNumber("Reset_Yaw_Count", reset_yaw_count);
            SmartDashboard::PutNumber("Reset_Displacement_Count", reset_displacement_count);
            SmartDashboard::PutBoolean("Yaw_Axis_Up",       yaw_axis_up);
            SmartDashboard::PutNumber( "Yaw_Axis", 			yaw_axis);

			SmartDashboard::PutBoolean( "IMU_Connected", imu->IsConnected());
			SmartDashboard::PutNumber("IMU_Yaw", imu->GetYaw());
			SmartDashboard::PutNumber("IMU_Pitch", imu->GetPitch());
			SmartDashboard::PutNumber("IMU_Roll", imu->GetRoll());
			SmartDashboard::PutNumber("IMU_CompassHeading", imu->GetCompassHeading());
			SmartDashboard::PutNumber("IMU_Update_Count", imu->GetUpdateCount());
			SmartDashboard::PutNumber("IMU_Byte_Count", imu->GetByteCount());

		    /* These functions are compatible w/the WPI Gyro Class */
			SmartDashboard::PutNumber(   "IMU_TotalYaw", 		imu->GetAngle());
			SmartDashboard::PutNumber(   "IMU_YawRateDPS",		imu->GetRate());

	        SmartDashboard::PutNumber("IMU_Accel_X", imu->GetWorldLinearAccelX());
			SmartDashboard::PutNumber("IMU_Accel_Y", imu->GetWorldLinearAccelY());
			SmartDashboard::PutBoolean("IMU_IsMoving", imu->IsMoving());
			SmartDashboard::PutNumber("IMU_Temp_C", imu->GetTempC());
            SmartDashboard::PutBoolean("IMU_IsCalibrating", imu->IsCalibrating());

            SmartDashboard::PutNumber("Velocity_X",       	imu->GetVelocityX() );
            SmartDashboard::PutNumber("Velocity_Y",       	imu->GetVelocityY() );
            SmartDashboard::PutNumber("Displacement_X",     imu->GetDisplacementX() );
            SmartDashboard::PutNumber("Displacement_Y",     imu->GetDisplacementY() );

            Wait(0.1);				// wait for a while
		}
	}

	/**
	 * Runs during test mode
	 */
	void Test() {

	}

};

START_ROBOT_CLASS(Robot);

File: roborio/java/navXMXPSimpleRobotExample/src/org/usfirst/frc/team2465/robot/Robot.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) FIRST 2008. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/
package org.usfirst.frc.team2465.robot;

import com.kauailabs.navx.frc.AHRS;

import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.SerialPort;
import edu.wpi.first.wpilibj.SampleRobot;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.livewindow.LiveWindow;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * The VM is configured to automatically run this class, and to call the
 * functions corresponding to each mode, as described in the SimpleRobot
 * documentation. If you change the name of this class or the package after
 * creating this project, you must also update the manifest file in the resource
 * directory.
 */
public class Robot extends SampleRobot {

    AHRS imu;			// This class can only be used w/the navX MXP.
    boolean first_iteration;

    public Robot() {

        try {

            // Use SerialPort.Port.kMXP if connecting navX MXP to the RoboRio MXP port
            // Use SerialPort.Port.kUSB if connecting navX MXP to the RoboRio USB port

            // You can add a second parameter to modify the 
            // update rate (in hz).  The minimum is 4.      
            // The maximum (and the default) is 60 on a navX MXP.
            // If you need to minimize CPU load, you can set it to a
            // lower value, as shown here, depending upon your needs.
            // The recommended update rate is 50Hz

            // The AHRS class provides access to advanced features on 
            // a navX MXP, including 9-axis headings
            // and magnetic disturbance detection.  This class also offers
            // access to altitude/barometric pressure data from a
            // navX MXP Aero.

            byte update_rate_hz = 50;
            imu = new AHRS(SerialPort.Port.kMXP, update_rate_hz);
        } catch( Exception ex ) {
            ex.printStackTrace();
        }
        if ( imu != null ) {
            LiveWindow.addSensor("IMU", "Gyro", imu);
        }
        first_iteration = true;
    }

    /**
     * This function is called once each time the robot enters autonomous mode.
     */
    public void autonomous() {

    }

    /**
     * This function is called once each time the robot enters operator control.
     */
    public void operatorControl() {

        AHRS.BoardYawAxis yaw_axis = new AHRS.BoardYawAxis();
        while (isOperatorControl() && isEnabled()) {

            // When calibration has completed, zero the yaw
            // Calibration is complete approximately 20 seconds
            // after the robot is powered on.  During calibration,
            // the robot should be still.

            boolean is_calibrating = imu.isCalibrating();
            if ( first_iteration && !is_calibrating ) {
                Timer.delay( 0.3 );
                imu.zeroYaw();
                first_iteration = false;
            }

            // Update the dashboard with status and orientation
            // data from the navX MXP

            imu.getBoardYawAxis(yaw_axis);  

            SmartDashboard.putBoolean("Yaw_Axis_Up",        yaw_axis.up);
            SmartDashboard.putNumber( "Yaw_Axis",           yaw_axis.board_axis);            
            
            boolean reset_yaw_button_pressed = DriverStation.getInstance().getStickButton(0,(byte)1);
            if ( reset_yaw_button_pressed ) {
                imu.zeroYaw();
            }
            
            SmartDashboard.putBoolean(  "IMU_Connected",        imu.isConnected());
            SmartDashboard.putBoolean(  "IMU_IsCalibrating",    imu.isCalibrating());
            SmartDashboard.putNumber(   "IMU_Yaw",              imu.getYaw());
            SmartDashboard.putNumber(   "IMU_Pitch",            imu.getPitch());
            SmartDashboard.putNumber(   "IMU_Roll",             imu.getRoll());
            SmartDashboard.putNumber(   "IMU_CompassHeading",   imu.getCompassHeading());
            SmartDashboard.putNumber(   "IMU_Update_Count",     imu.getUpdateCount());
            SmartDashboard.putNumber(   "IMU_Byte_Count",       imu.getByteCount());

            /* These functions are compatible w/the WPI Gyro Class */
            SmartDashboard.putNumber(   "IMU_TotalYaw", 		imu.getAngle());
            SmartDashboard.putNumber(   "IMU_YawRateDPS",		imu.getRate());

            SmartDashboard.putNumber(   "IMU_Accel_X",          imu.getWorldLinearAccelX());
            SmartDashboard.putNumber(   "IMU_Accel_Y",          imu.getWorldLinearAccelY());
            SmartDashboard.putBoolean(  "IMU_IsMoving",         imu.isMoving());
            SmartDashboard.putNumber(   "IMU_Temp_C",           imu.getTempC());

            Timer.delay(0.1);
        }
    }

    /**
     * This function is called once each time the robot enters test mode.
     */
    public void test() {

    }
}
